docs/analysis/RLPX_HANDSHAKE_AND_MESSAGE_ENCODING_ANALYSIS.md

```diff
@@ -0,0 +1,739 @@
+      messageData = new RawMessage(msgId, Bytes.wrap(messagePayload));
+    }
+  } else {
+    messageData = new RawMessage(msgId, Bytes.wrap(messagePayload));
+  }
+```
+
+**Snappy Compression** (`SnappyCompressor.java`):
+```java
+public byte[] compress(final byte[] uncompressed) {
+  try {
+    return Snappy.compress(uncompressed);
+  } catch (final IOException e) {
+    throw new FramingException("Snappy compression failed", e);
+  }
+}
+
+public byte[] decompress(final byte[] compressed) {
+  try {
+    return Snappy.uncompress(compressed);
+  } catch (final IOException e) {
+    throw new FramingException("Snappy decompression failed", e);
+  }
+}
+```
+
+**Key Features**:
+- **Xerial Snappy library** (same as Fukuii)
+- **Compression state** toggled via enableCompression()/disableCompression()
+- **Conditional fallback**: Treats as uncompressed only if compression never succeeded before
+- **No explicit size limit** in decompression (relies on Snappy library limits)
+
+### Comparison Matrix: Compression
+
+| Feature | Fukuii | Core-Geth | Besu | Compatible |
+|---------|--------|-----------|------|------------|
+| **Snappy Library** | Xerial | golang/snappy | Xerial | ✅ |
+| **Activation** | p2pVersion >= 5 | SetSnappy() call | enableCompression() | ✅ |
+| **Wire Protocol Compression** | YES (all messages) | YES (all messages) | YES (all messages) | ✅ |
+| **Max Decompressed Size** | 16MB | ~16MB (maxUint24) | No explicit limit | ⚠️ |
+| **Decompression Fallback** | Always (graceful) | No | Conditional | ⚠️ |
+| **Error on Decompress Fail** | Warn + continue | Hard error | Hard error after first success | ⚠️ |
+
+**Key Difference**: Fukuii has more robust fallback logic for handling compression edge cases.
+
+## Critical Implementation Differences
+
+### 1. Compression Fallback Strategy
+
+**Issue**: Some peers may advertise p2pVersion >= 5 but occasionally send uncompressed data due to bugs or protocol deviations.
+
+**Fukuii Approach**:
+```scala
+// Always attempt decompression, but fall back gracefully
+decompressData(frameData, frame).recoverWith { case ex =>
+  log.warn("COMPRESSION_FALLBACK: Treating as uncompressed")
+  Success(frameData)
+}
+```
+✅ **Most tolerant** - Always tries to continue communication
+
+**Core-Geth Approach**:
+```go
+// Return error if decompression fails
+data, err = snappy.Decode(c.snappyReadBuffer, data)
+if err != nil {
+  return code, nil, 0, err
+}
+```
+❌ **Strict** - Disconnects on decompression error
+
+**Besu Approach**:
+```java
+// Conditional fallback - only if compression never succeeded
+if (compressionSuccessful) {
+  throw e;  // Real error
+} else {
+  // Treat as uncompressed for compatibility
+}
+```
+⚠️ **Hybrid** - Strict after first successful decompression
+
+**Recommendation**: Fukuii's approach is most robust for handling real-world protocol deviations.
+
+### 2. Message Size Limits
+
+**Fukuii**:
+```scala
+val MaxDecompressedLength = 16777216  // 16MB exactly
```

Comment:
this is non-trivial. fukuii is targeting exactly 16,777,216 and geth is using 16,777,215 meaning all of our messages could potentially be 1 byte to large and rejected or we begin misinterpreting messages. adapt to core geths standards